pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,873 @@
+	instanceInformar informers.InstanceInfoInformer,
+	scLister storageListers.StorageClassLister,
+) *multishareReconciler {
+	recon := &multishareReconciler{
+		clientset: clientset,
+		cloud:     config.Cloud,
+		config:    config,
+		scLister:  scLister,
+	}
+
+	recon.shareLister = shareInformer.Lister()
+	recon.shareListerSynced = shareInformer.Informer().HasSynced
+
+	recon.instanceLister = instanceInformar.Lister()
+	recon.instanceListerSynced = instanceInformar.Informer().HasSynced
+
+	return recon
+}
+
+func (recon *multishareReconciler) Run(stopCh <-chan struct{}) {
+	klog.Infof("Starting multishare reconciler")
+	defer klog.Infof("Shutting down multishare reconciler")
+
+	informerSynced := []cache.InformerSynced{recon.shareListerSynced, recon.instanceListerSynced}
+	if !cache.WaitForCacheSync(stopCh, informerSynced...) {
+		klog.Errorf("Cannot sync caches")
+		return
+	}
+
+	go wait.Until(recon.reconcileWorker, 1*time.Minute, stopCh)
+
+	<-stopCh
+}
+
+func (recon *multishareReconciler) reconcileWorker() {
+	startTime := time.Now()
+
+	shares, err := recon.cloud.File.ListShares(context.TODO(), &file.ListFilter{Project: recon.cloud.Project, Location: "-", InstanceName: "-"})
+	if err != nil {
+		klog.Errorf("Reconciler Failed to list Shares: %v", err)
+	}
+	instances, err := recon.cloud.File.ListMultishareInstances(context.TODO(), &file.ListFilter{Project: recon.cloud.Project, Location: "-"})
+	if err != nil {
+		klog.Errorf("Reconciler Failed to list Instances: %v", err)
+	}
+	klog.V(6).Infof("found %d shares and %d instances", len(shares), len(instances))
+	// ops, err := recon.cloud.File.ListOps(context.TODO(), &file.ListFilter{Project: recon.cloud.Project, Location: "-"})
+	// if err != nil {
+	// 	klog.Errorf("Reconciler Failed to list ops: %v", err)
+	// }
+	instances, shares, instanceShares := recon.managedInstanceAndShare(instances, shares)
+
+	shareInfoMap := recon.createAndUpdateShareInfos(shares)
+
+	shareInfoList, err := recon.shareLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Filestore CSI driver cannot list ShareInfo objects: %v", err)
+		return
+	}
+	klog.V(6).Infof("listed out %d shareInfo objects", len(shareInfoList))
+	for _, shareInfo := range shareInfoList {
+		shareName := shareInfo.Name
+		if _, ok := shareInfoMap[shareName]; !ok {
+			shareInfoMap[shareName] = shareInfo
+		}
+	}
+
+	instanceInfoMap := recon.createAndUpdateInstanceInfos(instances, instanceShares)
+
+	instanceInfoList, err := recon.instanceLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Filestore CSI driver cannot list InstanceInfo objects: %v", err)
+		return
+	}
+	klog.V(6).Infof("listed out %d instanceInfo objects", len(instanceInfoList))
+	for _, instanceInfo := range instanceInfoList {
+		instanceURI := util.IINameToInstanceURI(instanceInfo.Name)
+		if _, ok := instanceInfoMap[instanceURI]; !ok {
+			instanceInfo, err := recon.maybeRemoveInstanceInfoFinalizer(instanceInfo)
```

Comment:
Please put some detailed comment, on how the instanceinfo is cleaned up after the finalizer is removed. Its not trivial

Reply:
done